stellargraph/connector/neo4j/graph.py

```diff
@@ -51,32 +71,139 @@ def nodes(self):
             The node IDs of all the nodes in the graph.
         """
         node_ids_query = f"""
-            MATCH (n)
-            RETURN n.ID as node_id
+            {self._match_node()}
+            RETURN node.ID as node_id
             """
 
         result = self.graph_db.run(node_ids_query)
         return [row["node_id"] for row in result.data()]
 
-    def node_features(self, node_ids):
+    def cache_all_nodes_in_memory(self, dtype="float32"):
+        """
+        Load all node IDs and features into memory from Neo4j so that subsequent method calls that
+        access node features can use the cached data instead of querying the database.
+
+        This method should be avoided for larger graphs.
+
+        Args:
+            dtype (str, optional): Data type of features
+
+        """
+        features = self._node_features_from_db(None)
+        self._nodes = convert.convert_nodes(
+            features, name="nodes", default_type=self._node_type, dtype=dtype,
+        )
+
+        # cache feature size too
+        self._node_feature_size = self._nodes.features_of_type(self._node_type).shape[1]
+
+    def _node_features_from_db(self, nodes):
+        return_node = f"""
+            WITH {{ID: node.ID, features: node.features}} AS node_data
+            RETURN node_data
+            """
+        if nodes is None:
+            feature_query = f"""
+                {self._match_node()}
+                {return_node}
+                """
+            result = self.graph_db.run(feature_query)
+            rows = result.data()
+            return IndexedArray(
+                np.array([node["node_data"]["features"] for node in rows]),
+                index=[node["node_data"]["ID"] for node in rows],
```

Comment:
Does iterating over this twice work correctly? I think for some databases, the result cursor is a single use object and can't be used again?

Reply:
Hmm right, the following seems to work without problem:
```python
query = f"""
MATCH (node:paper)
RETURN node.ID as id LIMIT 5
"""

result = neo4j_graphdb.run(query)
data = result.data()

for d in data:
    print(d)

for d in data:
    print(d)
```
```
{'id': 31336}
{'id': 1061127}
{'id': 1106406}
{'id': 13195}
{'id': 37879}
{'id': 31336}
{'id': 1061127}
{'id': 1106406}
{'id': 13195}
{'id': 37879}
```
and this code path should be running during `cache_all_nodes_in_memory`

Reply:
Oh, right, it's the `data()` method, that does the conversion to a list.